app/gui/src/providers/auth.ts

```diff
@@ -0,0 +1,301 @@
+    await signOut()
+
+    toastSuccess.show(getText('deleteUserSuccess'))
+
+    return true
+  }
+
+  const restoreUser = async () => {
+    await restoreUserMutation.mutateAsync()
+
+    toastSuccess.show(getText('restoreUserSuccess'))
+
+    return true
+  }
+
+  /**
+   * Update the user session data in the React Query cache.
+   * This only works for full user sessions.
+   * @deprecated Never use this function. Prefer particular functions like `setUsername` or `deleteUser`.
+   */
+  const setUser = (user: Partial<backendModule.User>) => {
+    const currentUser = queryClient.getQueryData(usersMeQueryOptions.queryKey)
+
+    if (currentUser != null && currentUser.type === UserSessionType.full) {
+      const currentUserData = currentUser.user
+      const nextUserData: backendModule.User = Object.assign(currentUserData, user)
+
+      queryClient.setQueryData(usersMeQueryOptions.queryKey, {
+        ...currentUser,
+        user: nextUserData,
+      })
+    }
+  }
+
+  const isUserMarkedForDeletion = () => userPromise.value.then((user) => !!user?.removeAt)
+
+  const isUserDeleted = async () => {
+    const user = await userPromise.value
+    if (user?.removeAt) {
+      const removeAtDate = new Date(user.removeAt)
+      const now = new Date()
+
+      return removeAtDate <= now
+    } else {
+      return false
+    }
+  }
+
+  const isUserSoftDeleted = async () => {
+    const user = await userPromise.value
+    if (user?.removeAt) {
+      const removeAtDate = new Date(user.removeAt)
+      const now = new Date()
+
+      return removeAtDate > now
+    } else {
+      return false
+    }
+  }
+
+  watchEffect(() => {
+    if (userData.value?.type === UserSessionType.full) {
+      sentry.setUser({
+        id: userData.value.user.userId,
+        email: userData.value.email,
+        username: userData.value.user.name,
+        // eslint-disable-next-line camelcase
+        ip_address: '{{auto}}',
+      })
+      onAuthenticated?.(userData.value.accessToken)
+    } else if (userData.value?.type === UserSessionType.partial) {
+      sentry.setUser({ email: userData.value.email })
+    }
+  })
+
+  gtagHooks.gtag('set', { platform: detect.platform(), architecture: detect.architecture() })
+  gtagHooks.gtagOpenCloseCallback(gtagEvent, 'open_app', 'close_app')
+
+  watchEffect(() => {
+    if (userData.value?.type === UserSessionType.full && userData.value.user.isEnsoTeamMember) {
+      setFeatureFlags(featureFlagsForInternalTesting())
+    }
+  })
+
+  const effectiveUserData = computed(() =>
+    userData.value?.type === UserSessionType.full && planOverride.value != null ?
+      { ...userData.value, user: { ...userData.value.user, plan: planOverride.value } }
+    : userData.value,
+  )
+
+  /**
+   * Check if given route is allowed for the current user.
+   *
+   * @returns boolean specifying if the route is allowed, or necessary redirect.
+   */
+  function routeGuard(route: RouteLocation) {
```

Comment:
`true | false | object` is a bit awkward. This seems like a situation for `Result`?